api/pkg/auth/auth.go

```diff
@@ -0,0 +1,240 @@
+package auth
+
+import (
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/markbates/goth"
+	"github.com/markbates/goth/gothic"
+	"github.com/tektoncd/hub/api/gen/log"
+	"github.com/tektoncd/hub/api/pkg/app"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"github.com/tektoncd/hub/api/pkg/token"
+	"gorm.io/gorm"
+)
+
+type service struct {
+	app.Service
+	api app.Config
+}
+
+type request struct {
+	db            *gorm.DB
+	log           *log.Logger
+	defaultScopes []string
+	jwtConfig     *app.JWTConfig
+}
+
+type Service interface {
+	AuthCallBack(res http.ResponseWriter, req *http.Request)
+
+	HubAuthenticate(res http.ResponseWriter, req *http.Request)
+}
+
+// New returns the resource service implementation.
+func New(api app.Config) Service {
+	return &service{
+		Service: api.Service("auth"),
+		api:     api,
+	}
+}
+
+func Authenticate(res http.ResponseWriter, req *http.Request) {
+	gothic.BeginAuthHandler(res, req)
+}
+
+func (s *service) AuthCallBack(res http.ResponseWriter, req *http.Request) {
+	ghUser, err := gothic.CompleteUserAuth(res, req)
+	if err != nil {
+		fmt.Fprintln(res, err)
+		return
+	}
+
+	r := request{
+		db:            s.DB(context.Background()),
+		log:           s.Logger(context.Background()),
+		defaultScopes: s.api.Data().Default.Scopes,
+		jwtConfig:     s.api.JWTConfig(),
+	}
+
+	var UI_URL = os.Getenv("UI_URL")
+	params := req.URL.Query()
+	if err = r.insertData(ghUser, params.Get("code")); err != nil {
```

Comment:
is `code` available with all provider?

Reply:
what is the primary info across all platform? mail id could be

Reply:
>  is code available with all provider?

yes

> what is the primary info across all platform? mail id could be

Yes we can have mail as primary for all the platforms but saying that we need to keep in mind for the users who are already logged in before we add that